hir/src/program/linker.rs

```diff
@@ -0,0 +1,594 @@
+    /// as memory management primitives. This function handles defining these prior to
+    /// linking the program.
+    fn populate_builtins(&mut self) {
+        // We provide three globals for managing the heap, based on the layout
+        // of the data segments and these globals.
+        let mut globals_offset = 0;
+        if let Some(last_segment) = self.program.segments.last() {
+            globals_offset = last_segment.offset() + last_segment.size();
+            // Ensure the start of the globals segment is word-aligned
+            let align_offset = globals_offset % 32;
+            if align_offset != 0 {
+                globals_offset += 32 - align_offset;
+            }
+        }
+        // Compute the start of the heap by finding the end of the globals segment, aligned to the nearest word boundary
+        let mut heap_base = globals_offset
+            .checked_add(
+                self.program
+                    .globals
+                    .size_in_bytes()
+                    .try_into()
+                    .expect("unable to allocate globals, unable to fit in linear memory"),
+            )
+            .expect("unable to allocate globals, not enough unreserved space available");
+        let align_offset = heap_base % 32;
+        if align_offset != 0 {
+            heap_base += 32 - align_offset;
+        }
+        let hp = heap_base.to_le_bytes();
+        // Initialize all 3 globals with the computed heap pointer
+        let heap_ptr_ty = Type::Ptr(Box::new(Type::U8));
+        self.program.globals.declare("HEAP_BASE".into(), heap_ptr_ty.clone(), Linkage::External, Some(hp.into())).expect("unable to declare HEAP_BASE, a conflicting global by that name was already defined");
+        self.program
+            .globals
+            .declare(
+                "HEAP_TOP".into(),
+                heap_ptr_ty.clone(),
+                Linkage::External,
+                Some(hp.into()),
+            )
+            .expect(
+                "unable to declare HEAP_TOP, a conflicting global by that name was already defined",
+            );
+        self.program
+            .globals
+            .declare(
+                "HEAP_END".into(),
+                heap_ptr_ty,
+                Linkage::External,
+                Some(hp.into()),
+            )
+            .expect(
+                "unable to declare HEAP_END, a conflicting global by that name was already defined",
+            );
+    }
+
+    /// If an executable is being linked, discover unused functions and garbage collect them.
+    ///
+    /// Once a function has been identified as dead and is collected, any transitive items it
+    /// references may also be collected if they are orphaned as a result of the collection.
+    ///
+    /// If a library is being linked, this function is a no-op, as it is not known what will
+    /// be needed at runtime once used in the context of a program.
+    fn garbage_collect(&mut self) {
+        // TODO:
```

Comment:
Is this on your immediate TODO list? If not, then we should create an issue for it.

Reply:
Good suggestion, I wasn't sure if I was going to get to it while I was working on the linker, but I think I'm going to defer this to a later time, it's not necessary for now.